app/assets/javascripts/angular/services/drag_selection.js

```diff
@@ -0,0 +1,110 @@
+angular.module('Prometheus.services').factory('DragSelection', ['$rootScope', function($rootScope) {
+  var ESCAPE_KEYCODE = 27;
+  var graph;
+  var drag;
+  var width;
+  var rectangle;
+
+  function pointAsDate(e) {
+    // Rounds to the nearest integer the x-axis value
+    // the mouse is hovering at.
+    return Math.floor(graph.x.invert(e.offsetX));
+  }
+
+  function onMouseup(datum, index) {
+    var pixelSelection = Math.abs(drag.stopPX - drag.startPX);
+    if (pixelSelection < 25) {
+      reset();
+      return;
+    }
+
+    drag.stopDt = pointAsDate(d3.event);
+    var windowAfterDrag = [
+      drag.startDt,
+      drag.stopDt
+    ].sort(compareNumbers);
+
+    graph.window.xMin = windowAfterDrag[0];
+    graph.window.xMax = windowAfterDrag[1];
+    graph.update();
+
+    var endTime = graph.window.xMax;
+    var range = graph.window.xMax - graph.window.xMin;
+    $rootScope.$broadcast('timeRangeRescale', {range: range, endTime: endTime * 1000});
+
+    reset();
+  }
+
+  function reset() {
+    d3.select(document).on("mousemove", null);
+    d3.select(document).on("mouseup", null);
+    d3.select(document).on("touchmove", null);
+    d3.select(document).on("touchend, touchcancel", null); // i think i can listen to both events like this..
+    drag = {};
+    rectangle.remove()
+  }
+
+  function compareNumbers(a, b) {
+    return a - b;
+  }
+
+  function onMousemove() {
+    var offset = drag.stopPX = d3.event.offsetX;
+    if (offset > (width - 1) || offset < 1) {
+      return
+    }
+    // Draw the rectangle
+    // javascript sort() by default sorts by string Unicode code points:
+    // [361, 100].sort() => [100, 361]
+    // [361, 99].sort() => [361, 99]
+    // have to use number-specific compare fn
+    var limits = [drag.startPX, d3.event.offsetX].sort(compareNumbers)
+    rectangle
+    .attr("fill", "steelblue")
+    .attr("x", limits[0])
+    .attr("width", limits[1]-limits[0]);
+  }
+
+  function onMousedown() {
+    rectangle = d3.select(this).append("rect")
+    .style("opacity", 0.5)
+    .attr("y", 0)
+    .attr("height", "100%");
+
+    drag.target = d3.event.target;
+    drag.startDt = pointAsDate(d3.event);
+    drag.y = d3.event.offsetY;
+    drag.startPX = d3.event.offsetX;
+    d3.event.preventDefault ? d3.event.preventDefault() : d3.event.returnValue = false;
+    d3.select(document).on("mousemove", onMousemove);
+    d3.select(document).on("mouseup", onMouseup);
+    d3.select(document).on("keyup", function() {
+      if (d3.event.keyCode === ESCAPE_KEYCODE) {
+        reset();
+      }
+    });
+    d3.select(document).on("touchmove", onMousemove);
+    d3.select(document).on("touchend, touchcancel", onMouseup); // i think i can listen to both events like this..
+    // d3.select(document).on("touchcancel", onMouseup);
```

Comment:
need to get my hands on an ipad at work, would love to test using touch for zooming :)